consensus/replay_test.go

```diff
@@ -56,43 +60,85 @@ var testLog = `{"time":"2016-04-03T11:23:54.387Z","msg":[3,{"duration":972835254
 {"time":"2016-04-03T11:23:54.392Z","msg":[2,{"msg":[20,{"ValidatorIndex":0,"Vote":{"height":1,"round":0,"type":2,"block_hash":"4291966B8A9DFBA00AEC7C700F2718E61DF4331D","block_parts_header":{"total":1,"hash":"3BA1E90CB868DA6B4FD7F3589826EC461E9EB4EF"},"signature":"39147DA595F08B73CF8C899967C8403B5872FD9042FFA4E239159E0B6C5D9665C9CA81D766EACA2AE658872F94C2FCD1E34BF51859CD5B274DA8512BACE4B50D"}}],"peer_key":""}]}
 `
 
-func TestReplayCatchup(t *testing.T) {
+// map lines in the above wal to privVal step
+var mapPrivValStep = map[int]int8{
+	0: 0,
+	1: 0,
+	2: 1,
+	3: 1,
+	4: 1,
+	5: 2,
+	6: 2,
+	7: 3,
+}
+
+func writeWAL(log string) string {
+	fmt.Println("writing", log)
 	// write the needed wal to file
 	f, err := ioutil.TempFile(os.TempDir(), "replay_test_")
 	if err != nil {
 		panic(err)
 	}
-	name := f.Name()
-	_, err = f.WriteString(testLog)
+
+	_, err = f.WriteString(log)
 	if err != nil {
 		panic(err)
 	}
+	name := f.Name()
 	f.Close()
+	return name
+}
+
+func waitForBlock(newBlockCh chan interface{}) {
+	after := time.After(time.Second * 10)
+	select {
+	case <-newBlockCh:
+	case <-after:
+		panic("Timed out waiting for new block")
+	}
+}
+
+func runReplayTest(t *testing.T, cs *ConsensusState, fileName string, newBlockCh chan interface{}) {
+	// open wal and run catchup messages
+	openWAL(t, cs, fileName)
+	go cs.timeoutRoutine()
+	if err := cs.catchupReplay(cs.Height); err != nil {
+		panic(Fmt("Error on catchup replay %v", err))
+	}
+	go cs.receiveRoutine(0)
+	// wait to make a new block
+	waitForBlock(newBlockCh)
+	cs.QuitService.OnStop()
+}
+
+func setupReplayTest(nLines int, crashAfter bool) (*ConsensusState, chan interface{}, string, string) {
+	fmt.Println("-------------------------------------")
+	log.Notice(Fmt("Starting replay test of %d lines of WAL (crash before write)", nLines))
+
+	lineStep := nLines
+	if crashAfter {
+		lineStep -= 1
+	}
+
+	split := strings.Split(testLog, "\n")
+	lastMsg := split[nLines]
+
+	// we write those lines up to (not including) one with the signature
+	fileName := writeWAL(strings.Join(split[:nLines], "\n") + "\n")
 
 	cs := fixedConsensusState()
+	cs.QuitService.OnStart()
```

Comment:
actually, i don't think we should even specify .QuitService. either, in case the service struct changes, no?